File: Arrays & String/Problem6 - 151.java
public class Main {

    /* Optimal Approach - using extra space */
    public static String reverseWords(String s) {
        int n = s.length();
        StringBuilder reverseWord = new StringBuilder();
        StringBuilder temp = new StringBuilder();

        int i = n - 1;
        while (i >= 0) {

            // take the word 
            while (i >= 0 && s.charAt(i) != ' ') {
                temp.insert(0, s.charAt(i));
                i--;
            }

            // remove trailing spaces 
            while (i >= 0 && s.charAt(i) == ' ') i--;

            if (temp.length() > 0) {
                if (i >= 0) temp.insert(0, ' '); // we have further word in future, so add space
                reverseWord.append(temp);
                temp.setLength(0); // clear temp StringBuilder
            }
        }
        return reverseWord.toString();
    }

    public static void main(String[] args) {
        String s = "  hello world  ";
        System.out.println(reverseWords(s)); // Output: "world hello"
    }
}

File: Arrays & String/Problem6 - 151.cpp
#include <bits/stdc++.h>
using namespace std;

/* Optimal Approach - using extra space */
string reverseWords(string s) {
    int n = s.size();
    string reverseWord = "";
    string temp = "";

    int i = n - 1;
    while (i >= 0) {

        // take the word 
        while (i >= 0 && s[i] != ' ') {
            temp = s[i] + temp;
            i--;
        }

        // remove tralling zero 
        while (i >= 0 && s[i] == ' ') i--;

        if (temp.size() > 0) {
            if (i >= 0)  temp += ' '; // we have further word in future
            // so take space 
            reverseWord += temp;
            temp = "";
        }
    }
    return reverseWord;
    /*  TC-O(n) SC-O(n) */
}
 